src/main/java/org/socialhistoryservices/delivery/reservation/controller/ReservationController.java

```diff
@@ -382,45 +435,119 @@ protected void autoPrint(final Reservation res) {
 
     /**
      * Mass delete reservations.
-     * @param req The HTTP request object.
+     *
+     * @param req     The HTTP request object.
      * @param checked The reservations marked for deletion.
      * @return The view to resolve.
      */
     @RequestMapping(value = "/batchprocess",
-                    method = RequestMethod.POST,
-                    params = "delete")
+            method = RequestMethod.POST,
+            params = "delete")
     @PreAuthorize("hasRole('ROLE_RESERVATION_DELETE')")
     public String batchProcessDelete(HttpServletRequest req,
-                                     @RequestParam(required=false) List<String>
+                                     @RequestParam(required = false) List<String>
                                              checked) {
-
         // Delete all the provided reservations
         if (checked != null) {
+            ArrayList<Integer> requestIds = new ArrayList<>();
+
+            for (BulkActionIds bulkActionIds : getIdsFromBulk(checked)) {
+                if (!requestIds.contains(bulkActionIds.getRequestId())) {
+                    requestIds.add(bulkActionIds.getRequestId());
+                }
+            }
+
+            // loop each reservation
+            for (Integer resId : requestIds) {
+                changeStatusOfHoldingsInReservation(resId, Holding.Status.RESERVED, Holding.Status.AVAILABLE);
+            }
 
             for (BulkActionIds bulkActionIds : getIdsFromBulk(checked)) {
                 delete(bulkActionIds.getRequestId());
             }
         }
+
         String qs = req.getQueryString() != null ?
-                    "?" + req.getQueryString() : "";
+                "?" + req.getQueryString() : "";
         return "redirect:/reservation/" + qs;
     }
 
+    private void changeStatusOfHoldingsInReservation(Integer reservationId, Holding.Status StatusFrom, Holding.Status StatusTo) {
```

Comment:
This method should be made more general, so you don't repeat yourself. And it shouldn't happen on the controller level.